File: tests/Tests.CSharp/FrameDeltaCrossVerifyTests.cs
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using Xunit;
using Zeta.Core.CSharp;

namespace Zeta.Tests.CSharp;

/// <summary>
/// FrameDelta cross-language agreement — the C# oracle replays the shared seed
/// (<c>src/Core.TypeScript/frame-delta/golden-vectors.json</c>) and asserts identical compose / inverse /
/// between / apply / magnitude / distance results. Passing == agreeing with the F# oracle on the
/// frame-offset transformation group.
/// </summary>
public class FrameDeltaCrossVerifyTests
{
    private static string RepoRoot()
    {
        var dir = new DirectoryInfo(Path.GetDirectoryName(typeof(FrameDeltaCrossVerifyTests).Assembly.Location)!);
        while (dir is not null && !File.Exists(Path.Join(dir.FullName, "Zeta.sln")))
        {
            dir = dir.Parent;
        }

        return dir?.FullName ?? throw new DirectoryNotFoundException("Could not locate repo root (Zeta.sln).");
    }

    private static Dictionary<string, long> Map(JsonElement obj) =>
        obj.EnumerateObject().ToDictionary(p => p.Name, p => p.Value.GetInt64(), System.StringComparer.Ordinal);

    private static JsonElement Seed()
    {
        var path = Path.Join(RepoRoot(), "src", "Core.TypeScript", "frame-delta", "golden-vectors.json");
        Assert.True(File.Exists(path), $"seed not found: {path}");
        using var doc = JsonDocument.Parse(File.ReadAllText(path));
        return doc.RootElement.Clone();
    }

    private static void AssertMapEqual(Dictionary<string, long> expected, IReadOnlyDictionary<string, long> actual)
    {
        Assert.Equal(expected.Count, actual.Count);
        foreach (var kv in expected)
        {
            Assert.True(actual.TryGetValue(kv.Key, out var v) && v == kv.Value, $"key {kv.Key}");
        }
    }

    [Fact]
    public void CSharpFrameDeltaAgreesWithSeed()
    {
        var seed = Seed();

        foreach (var v in seed.GetProperty("compose").EnumerateArray())
        {
            AssertMapEqual(Map(v.GetProperty("result")), FrameDelta.Compose(Map(v.GetProperty("a")), Map(v.GetProperty("b"))));
        }

        foreach (var v in seed.GetProperty("inverse").EnumerateArray())
        {
            AssertMapEqual(Map(v.GetProperty("result")), FrameDelta.Inverse(Map(v.GetProperty("d"))));
        }

        foreach (var v in seed.GetProperty("between").EnumerateArray())
        {
            AssertMapEqual(Map(v.GetProperty("result")), FrameDelta.Between(Map(v.GetProperty("from")), Map(v.GetProperty("to"))));
        }

        foreach (var v in seed.GetProperty("apply").EnumerateArray())
        {
            AssertMapEqual(Map(v.GetProperty("result")), FrameDelta.Apply(Map(v.GetProperty("delta")), Map(v.GetProperty("frame"))));
        }

        foreach (var v in seed.GetProperty("magnitude").EnumerateArray())
        {
            Assert.Equal(v.GetProperty("result").GetInt64(), FrameDelta.Magnitude(Map(v.GetProperty("d"))));
        }

        foreach (var v in seed.GetProperty("distance").EnumerateArray())
        {
            Assert.Equal(v.GetProperty("result").GetInt64(), FrameDelta.Distance(Map(v.GetProperty("from")), Map(v.GetProperty("to"))));
        }
    }
}

File: tests/Tests.FSharp/FrameDelta.CrossVerify.Tests.fs
module Zeta.Tests.FrameDeltaCrossVerifyTests

open System.IO
open System.Text.Json
open global.Xunit
open Zeta.Core

module FD = Zeta.Core.FrameDelta
module TF = Zeta.Core.TravelerFrame

// ═══════════════════════════════════════════════════════════════════
// FrameDelta cross-language agreement — the F# oracle replays the shared seed
// (src/Core.TypeScript/frame-delta/golden-vectors.json) that the C# oracle (and, in turn, TS/Rust) also
// verify. Both passing == agreement on the frame-offset transformation group (the 4-lang leg).
// ═══════════════════════════════════════════════════════════════════

let private repoRoot () =
    let mutable dir = DirectoryInfo(Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location))
    while not (isNull dir) && not (File.Exists(Path.Join(dir.FullName, "Zeta.sln"))) do
        dir <- dir.Parent
    if isNull dir then failwith "Could not locate repo root (Zeta.sln)." else dir.FullName

let private toMap (e: JsonElement) : Map<string, int64> =
    [ for p in e.EnumerateObject() -> p.Name, p.Value.GetInt64() ] |> Map.ofList

let private toDelta (e: JsonElement) : FD.Delta = { FD.Shifts = toMap e }
let private toFrame (e: JsonElement) : TF.Frame =
    { TF.Coords = toMap e |> Map.map (fun _ v -> Versionstamp.ofInt64 v) }
let private frameMap (f: TF.Frame) : Map<string, int64> = f.Coords |> Map.map (fun _ vs -> vs.Version)

let private seed () =
    let path = Path.Join(repoRoot (), "src", "Core.TypeScript", "frame-delta", "golden-vectors.json")
    Assert.True(File.Exists path, sprintf "seed not found: %s" path)
    JsonDocument.Parse(File.ReadAllText path)

[<Fact>]
let ``F# FrameDelta agrees with the shared golden seed`` () =
    use doc = seed ()
    let root = doc.RootElement
    let section (name: string) : JsonElement[] = root.GetProperty(name).EnumerateArray() |> Seq.toArray

    for v in section "compose" do
        Assert.Equal<Map<string, int64>>(toMap (v.GetProperty "result"), (FD.compose (toDelta (v.GetProperty "a")) (toDelta (v.GetProperty "b"))).Shifts)

    for v in section "inverse" do
        Assert.Equal<Map<string, int64>>(toMap (v.GetProperty "result"), (FD.inverse (toDelta (v.GetProperty "d"))).Shifts)

    for v in section "between" do
        Assert.Equal<Map<string, int64>>(toMap (v.GetProperty "result"), (FD.between (toFrame (v.GetProperty "from")) (toFrame (v.GetProperty "to"))).Shifts)

    for v in section "apply" do
        Assert.Equal<Map<string, int64>>(toMap (v.GetProperty "result"), frameMap (FD.apply (toDelta (v.GetProperty "delta")) (toFrame (v.GetProperty "frame"))))

    for v in section "magnitude" do
        Assert.Equal<int64>((v.GetProperty "result").GetInt64(), FD.magnitude (toDelta (v.GetProperty "d")))

    for v in section "distance" do
        Assert.Equal<int64>((v.GetProperty "result").GetInt64(), FD.distance (toFrame (v.GetProperty "from")) (toFrame (v.GetProperty "to")))

    // homeostat leg (order-independent aggregation) across the oracle: folding the deltas in any order
    // gives the same total.
    for v in section "aggregate" do
        let deltas = [ for d in (v.GetProperty "deltas").EnumerateArray() -> toDelta d ]
        let total = toMap (v.GetProperty "total")
        let fold order = (List.fold FD.compose FD.identity order).Shifts
        Assert.Equal<Map<string, int64>>(total, fold deltas)
        Assert.Equal<Map<string, int64>>(total, fold (List.rev deltas))
